process/block/preprocess/transactions.go

```diff
@@ -630,80 +876,127 @@ func (txs *transactions) createAndProcessMiniBlock(
 			txs.gasHandler.RemoveGasConsumed([][]byte{txHash})
 			txs.gasHandler.RemoveGasRefunded([][]byte{txHash})
 
-			gasConsumedByMiniBlockInSenderShard = oldGasConsumedByMiniBlockInSenderShard
-			gasConsumedByMiniBlockInReceiverShard = oldGasConsumedByMiniBlockInReceiverShard
-
+			gasConsumedByMiniBlocksInSenderShard = oldGasConsumedByMiniBlocksInSenderShard
+			mapGasConsumedByMiniBlockInReceiverShard[receiverShardID] = oldGasConsumedByMiniBlockInReceiverShard
+			totalGasConsumedInSelfShard = oldTotalGasConsumedInSelfShard
 			continue
 		}
 
-		gasRefunded := txs.gasHandler.GasRefunded(txHash)
-		gasConsumedByMiniBlockInReceiverShard -= gasRefunded
-		if senderShardId == receiverShardId {
-			gasConsumedByMiniBlockInSenderShard -= gasRefunded
-		}
+		senderAddressToSkip = []byte("")
 
-		if !errors.Is(err, process.ErrFailedTransaction) {
-			miniBlock.TxHashes = append(miniBlock.TxHashes, txHash)
+		gasRefunded := txs.gasHandler.GasRefunded(txHash)
+		mapGasConsumedByMiniBlockInReceiverShard[receiverShardID] -= gasRefunded
+		if senderShardID == receiverShardID {
+			gasConsumedByMiniBlocksInSenderShard -= gasRefunded
+			totalGasConsumedInSelfShard -= gasRefunded
 		}
-		addedTxs++
 
-		if addedTxs >= spaceRemained { // max transactions count in one block was reached
-			log.Debug("max txs accepted in one block is reached",
-				"num added txs", len(miniBlock.TxHashes),
-				"total txs", len(orderedTxs),
-			)
+		if errors.Is(err, process.ErrFailedTransaction) {
+			if !firstInvalidTxFound {
+				firstInvalidTxFound = true
+				txs.blockSizeComputation.AddNumMiniBlocks(1)
+			}
 
-			log.Debug("mini block info",
-				"gas consumed in sender shard", gasConsumedByMiniBlockInSenderShard,
-				"gas consumed in receiver shard", gasConsumedByMiniBlockInReceiverShard,
-				"gas consumed in self shard", txs.gasHandler.TotalGasConsumed(),
-				"txs ordered", len(orderedTxs),
-				"txs added", len(miniBlock.TxHashes))
+			txs.blockSizeComputation.AddNumTxs(1)
+			continue
+		}
 
-			return miniBlock, nil
+		if _, ok := mapMiniBlocks[receiverShardID]; !ok {
+			mapMiniBlocks[receiverShardID] = txs.createEmptyMiniBlock(senderShardID, receiverShardID, block.TxBlock)
+			txs.blockSizeComputation.AddNumMiniBlocks(1)
```

Comment:
you could create all the miniblocks, one for every shard. There should be no problem to have space for ONE MiniBlock per Shard from transaction preprocessor. And finaly when creating the slice of miniblocks from map, you add only those miniblocks which have txs.

Reply:
I want to check very strictly the space for each mini block and each tx, and I think it is not a problem with this approach as it is called only once for each different receiverShardID

Reply:
the code would be simpler.

Reply:
Refactored